### PR TITLE
Coordinated update with GraphQL upgrades in moesifmiddleware-go

### PR DESCRIPTION
**Pull request by bakennedy**
DRY up api endpoint response header handling for config and governance rules eTags, and use callback functions for async response value handling

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -110,7 +110,9 @@ type API interface {
 	// templated with individual user and company info from the /v1/config endpoint
 	GetGovernanceRules() (GovernanceRulesResponse, error)
 
-	GetETag() string
+	// SetEventsHeaderCallback sets a Moesif API response header value and callback
+	// function which is called with
+	SetEventsHeaderCallback(string, func(string))
```
Comment (bakennedy) [inline api.go:115]: This change is the focus of the eTag update fix

File: client.go [hunk 2]
```diff
@@ -22,47 +23,30 @@ import (
...
-		c.rulesETag = eTag[0]
-	}
+	cancel               func()
+	ctx                  context.Context
+	ch                   chan []*models.EventModel
+	chUser               chan []*models.UserModel
+	chCompany            chan []*models.CompanyModel
+	flush                chan chan struct{}
+	interval             time.Duration
+	timeout              time.Duration
```
Comment (praves77) [inline client.go:33]: we no longer need responseETag or rulesETag?

File: client.go [hunk 3]
```diff
@@ -468,3 +453,20 @@ func (c *Client) start() {
...
+// SetEventsHeaderCallback takes a response header name and callback function
+// on SendDataToMoesif, the callback headers are read from the response and
+// passed to the callback functions
+func (c *Client) SetEventsHeaderCallback(header string, callback func(string)) {
+	c.eventHeaderCallbacks[header] = callback
+}
+
+// notify iterates over event header callbacks, looks up the header value, and calls back
+// if a non empty value is found
+func (c *Client) notify(headers http.Header) {
```
Comment (bakennedy) [inline client.go:466]: eTag updates are pushed out from all responses here and processed by the receiver if necessary

**Comment by praves77** [comment]
Looks fine to me except a few questions.
Is your IDE using tab not spaces or otherwise? There are few format related changes.

**Review by praves77 — APPROVED** [review]

File: client.go (hunk 2 above)
Comment (bakennedy) [reply client.go:33]: Yep! previously concurrent goroutines just raced to overwrite those variables and eventually hopefully converged on the correct latest value, but now callbacks handle each response e-tag individually so no need to unsynchronized global state

**Comment by bakennedy** [comment]
> Looks fine to me except a few questions. Is your IDE using tab not spaces or otherwise? There are few format related changes.

In this case, I think it's just `go fmt` applying automatically on save and adjusting struct spacing to follow the go format standard.  Still using spaces for indentation

**Pull request merged**